wiring/src/spark_wiring_ble.cpp

```diff
@@ -883,10 +885,75 @@ class BleServiceImpl {
         return endHandle_;
     }
 
+    BleConnectionHandle& connHandle() {
+        return connHandle_;
+    }
+
+    bool& characteristicsDiscovered() {
+        return characteristicsDiscovered_;
+    }
+
+    Vector<BleCharacteristic>& characteristics() {
+        return characteristics_;
+    }
+
+    int discoverAllCharacteristics(BleService& service) {
+        LOG(TRACE, "Start discovering characteristics.");
+        hal_ble_svc_t halService;
+        halService.size = sizeof(hal_ble_svc_t);
+        halService.start_handle = service.impl()->startHandle();
+        halService.end_handle = service.impl()->endHandle();
+        CHECK(hal_ble_gatt_client_discover_characteristics(service.impl()->connHandle(), &halService, onCharacteristicsDiscovered, &service, nullptr));
+        for (auto& characteristic : service.impl()->characteristics()) {
+            // Read the user description string if presented.
+            if (characteristic.impl()->attrHandles().user_desc_handle != BLE_INVALID_ATTR_HANDLE) {
+                char desc[BLE_MAX_DESC_LEN] = {};
+                size_t len = hal_ble_gatt_client_read(service.impl()->connHandle(), characteristic.impl()->attrHandles().user_desc_handle, (uint8_t*)desc, sizeof(desc) - 1, nullptr);
+                if (len > 0) {
+                    desc[len] = '\0';
+                    characteristic.impl()->description() = desc;
+                    LOG_DEBUG(TRACE, "User description: %s.", desc);
+                }
+            }
+        }
+        return SYSTEM_ERROR_NONE;
+    }
+
 private:
+    BleConnectionHandle connHandle_; // For peer service
     BleUuid uuid_;
     BleAttributeHandle startHandle_;
     BleAttributeHandle endHandle_;
+    bool characteristicsDiscovered_;
+    Vector<BleCharacteristic> characteristics_;
+
+    static void onCharacteristicsDiscovered(const hal_ble_char_discovered_evt_t* event, void* context) {
```

Comment:
Same here. Please refer to: https://github.com/particle-iot/device-os/pull/2203/files#r493623290